### PR TITLE
Move TensorOptions to ATen/core

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#22020 Move TensorOptions to ATen/core**

In preparation for adding DimnameList (which is in ATen) to TensorOptions.
In the long term we'll move DimnameList to c10 because that is important
for named tensor performance (TensorImpl holds a DimnameList but right
now there is one layer of indirection to support it). However, for now,
since TensorOptions isn't used by c10, move it back into ATen/core.

Test Plan:
- wait for ci

gh-metadata: pytorch pytorch 22020 gh/zou3519/57/head

Differential Revision: [D15920763](https://our.internmc.facebook.com/intern/diff/D15920763)

### CODE REVIEW

**Comment by gchanan** [comment]
are there BC concerns with this?  Or does `c10::TensorOptions` still "work" (for external users).

**Comment by zou3519** [comment]
> are there BC concerns with this? Or does `c10::TensorOptions` still "work" (for external users).

c10::TensorOptions will no longer "work" for external users, so yes, there are BC concerns. However, there are two things that alleviate this concern:
1. TensorOptions is imported in ATen/ATen.h so the hope is that users don't directly import `c10/core/TensorOptions.h`.
2. I don't think many users will be creating TensorOptions manually: its use case as I understand it it is to make it easier to pass arguments into tensor constructors.

I'm not sure how to completely get around the BC concerns, though; we can't import `ATen/TensorOptions.h` from c10.

**Comment by gchanan** [comment]
1.  sure, I was worried about the C++ symbol, not the includes.
2.  Some people probably use them manually.

Do we not expose every symbol in aten in the c10 namespace somewhere?  Is it only in the other direction?

**Comment by zou3519** [comment]
> Do we not expose every symbol in aten in the c10 namespace somewhere? Is it only in the other direction?

I had to change usages of c10::TensorOptions to at::TensorOptions. Based on that experience, we definitely do not expose every symbol in aten in the c10 namespace.

In the other direction, I'm not sure that we expose every symbol in the c10  namespace in the ATen namespace. I don't see a general "import everything into ATen" file so figuring this out would require auditing c10 symbols on a case-by-case basis.

**Comment by smessmer** [comment]
Every `c10::xxx` is also exposed as `at::xxx`. Just keep it `c10::TensorOptions` and you can still use it as `at::TensorOptions` if you prefer that.

**Comment by zou3519** [comment]
> Every `c10::xxx` is also exposed as `at::xxx`. Just keep it `c10::TensorOptions` and you can still use it as `at::TensorOptions` if you prefer that.

Thanks for the confirmation. I've kept it as `c10::TensorOptions` as requested.

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/core/TensorOptions.h [hunk 1]
```diff
@@ -614,4 +614,10 @@ inline DeviceType computeDeviceType(TensorTypeId tid) {
   }
 }
 
+} // namespace at
+
+namespace c10 {
+
+using TensorOptions = at::TensorOptions;
```
Comment (jerryzh168) [inline aten/src/ATen/core/TensorOptions.h]: why adding new things to c10 namespace while we are in at?

Comment (zou3519) [reply aten/src/ATen/core/TensorOptions.h]: It's to keep backwards compatibility with users of the c++ API who may be referring to TensorOptions as `c10::TensorOptions`.

Comment (smessmer) [inline aten/src/ATen/core/TensorOptions.h]: Let's instead not change the namespace and keep TensorOptions in the c10:: namespace. Everything in the c10 namespace is automatically available in aten:: as well.

**Review by smessmer — CHANGES_REQUESTED** [review]
Generally looks good but many of the changes are unnecessary if you just keep TensorOptions in the c10:: namespace and only move the file to the ATen folder.

**Comment by zou3519** [comment]
@smessmer I've made the requested changes, please let me know how it looks now.

**Review by smessmer — APPROVED** [review]
looks good

**Pull request closed without merging**